### PR TITLE
Allow subdirectories to be tested by x.py test

### LINKED ISSUE #60718: x.py test should be able to test subfolders
It would be very useful if `./x.py test src/test/ui/subfolder` would run every UI test in `src/test/ui/subfolder` (at the moment it runs every single UI test). `./x.py test` currently works for [individual files](https://github.com/rust-lang/rust/issues/48483) and for entire test suites, but not directories.

### PR DESCRIPTION
**Pull request by varkor**
Fixes https://github.com/rust-lang/rust/issues/60718.

As far as I can tell, multiple `--test-args` flags are ignored (only the first is respected), so if you specify a subdirectory, you won't also be able to filter using `--test-args`. If you don't specify a subdirectory, `--test-args` will continue working as usual, so this is strictly an improvement on the current state of affairs.

### CODE REVIEW

**Comment by Centril** [comment]
@varkor Can you follow up with a PR to the rustc guide?

File: src/bootstrap/test.rs [hunk 1]
```diff
@@ -1184,8 +1184,13 @@ impl Step for Compiletest {
                     Err(_) => p,
                 }
             })
-            .filter(|p| p.starts_with(suite_path) && p.is_file())
-            .map(|p| p.strip_prefix(suite_path).unwrap().to_str().unwrap())
+            .filter(|p| p.starts_with(suite_path) && (p.is_dir() || p.is_file()))
+            .filter_map(|p| {
+                match p.strip_prefix(suite_path).ok().and_then(|p| p.to_str()) {
+                    Some(s) if s != "" => Some(s),
```
Comment (lzutao) [inline src/bootstrap/test.rs]: Is `str::is_empty` be more clear here?

**Comment by alexcrichton** [comment]
r? @Mark-Simulacrum 

This is pretty far out of cache for me so curious if you remember this more?

File: src/bootstrap/test.rs [hunk 2]
```diff
@@ -1184,8 +1184,13 @@ impl Step for Compiletest {
...
+                    _ => None,
```
Comment (Mark-Simulacrum) [inline src/bootstrap/test.rs]: It's unclear to me why we've switched to filter_map and None here -- could you write up a comment on the `None` branch?

**Comment by Mark-Simulacrum** [comment]
I would like to do another review after this one but seems reasonable (I don't honestly think this is really too viable to review thoroughly; if it passes CI and works in practice then that's probably good enough).

File: src/bootstrap/test.rs (hunk 2 above)
Comment (varkor) [reply src/bootstrap/test.rs]: I've added a comment with explanation.

File: src/bootstrap/test.rs (hunk 1 above)
Comment (varkor) [reply src/bootstrap/test.rs]: I don't think this is any clearer in this instance.

**Review by Mark-Simulacrum — APPROVED** [review]

**Pull request merged**